src/main/java/frc/robot/subsystems/targetvision/TargetVision.kt

```diff
@@ -1,61 +1,203 @@
 package frc.robot.subsystems.targetvision
 
 
-import edu.wpi.first.math.geometry.Rotation2d
+import edu.wpi.first.math.Num
 import edu.wpi.first.math.geometry.Translation2d
 import edu.wpi.first.math.util.Units
-import edu.wpi.first.math.util.Units.inchesToMeters
+import edu.wpi.first.wpilibj.DriverStation
 import edu.wpi.first.wpilibj2.command.Subsystem
-import frc.robot.utils.LimelightHelpers.LimelightTarget_Retro
+import frc.robot.utils.LimelightHelpers.LimelightTarget_Detector
 import org.littletonrobotics.junction.Logger
+import kotlin.math.abs
 import kotlin.math.tan
+import kotlin.math.pow
+import edu.wpi.first.math.Vector
+
+
+object TargetVision:  Subsystem {
 
-object TargetVision : Subsystem {
     private val io: TargetVisionIO = Limelight
-    private val inputs: TargetVisionIO.Inputs = TargetVisionIO.Inputs()
+    private val inputs: TargetVisionIO.TargetVisionIOInputs = TargetVisionIO.TargetVisionIOInputs()
     private const val SAMPLE_NUM = 8
 
+
+    private fun getDistance(target: LimelightTarget_Detector): Double{
+        val angleToBucket = Units.degreesToRadians(CAMERA_PITCH + target.ty)
+        return (BUCKET_HEIGHT - LIMELIGHT_HEIGHT) / tan(angleToBucket)
+    }
+
+    //ascending
+    private val targetsbyDistance: List<Pair<LimelightTarget_Detector, Double>>
+        get() {
+            return inputs.targets.map {target: LimelightTarget_Detector -> Pair(target, getDistance(target))}.sortedBy {it.second}
+        }
+    val hasTargets: Boolean
+        get(){
+            return inputs.hasTargets
+        }
+    val curTime: Double
+        get() {return inputs.lastTimeStampMS}
+
+
+
     override fun periodic() {
-        io.updateInputs(inputs)
-        Logger.getInstance().processInputs("Target Vision", inputs)
+        Logger.getInstance().processInputs("Vision", inputs)
 
-        if (inputs.targets.isNotEmpty()) {
-            addMeasurement(takeMeasurement(targetsByDistance[0].first))
+        for (target in targetsbyDistance) {
+            if(target.first.classID == (OpposingAlliance.id)){
+                
+                addMeasurement(takeMeasurement(target.first))
+            }
         }
+
+        io.updateInputs(inputs)
     }
 
-    // in ascending order
-    private val targetsByDistance: List<Pair<LimelightTarget_Retro, Double>>
-        get() {
-            return inputs.targets.map { target: LimelightTarget_Retro -> Pair(target, distanceToTarget(target)) }
-                .sortedBy { it.second }
+   data class Measurement(val timestamp: Double, val pose: LimelightTarget_Detector)
```

Comment:
What is even going on here? Why is there an inner class called `Sample` on `Measurement` when the words mean the same thing? Why is the `LimelightTarget_Detector` property called `pose`? Why does `takeMeasurement` take `target` as a parameter but use `inputs` from the outer class to get the timestamp? And why does it assign `target` to the `pose` property?

Reply:
pose thing again comes from an old version where it was a measurement that stored a pose. Sample stores a translation2d and measurement stores a ll target. we use sample for turret command stuff so we dont have to do the translation there, and measurement for most things here. all of the other questions are do to the wierd naming of pose, which is quite misleading